File: cmd/server/external_source_name.go
package main

import (
	"fmt"
	"strings"
	"unicode/utf8"
)

const maxExternalSourceNameCodePoints = 64

// validateExternalSourceName normalizes and validates names supplied through
// either source API. Database writes must use the returned value.
func validateExternalSourceName(raw string) (string, error) {
	name := strings.TrimSpace(raw)
	if name == "" {
		return "", fmt.Errorf("name is required")
	}
	if !utf8.ValidString(name) {
		return "", fmt.Errorf("name must be valid UTF-8")
	}
	if utf8.RuneCountInString(name) > maxExternalSourceNameCodePoints {
		return "", fmt.Errorf("name is too long (max %d characters)", maxExternalSourceNameCodePoints)
	}
	return name, nil
}

// normalizeRemoteExternalSourceName is intentionally separate from user-input
// validation: remote metadata can be shortened, while user input is rejected.
func normalizeRemoteExternalSourceName(raw string) (string, bool) {
	name := strings.TrimSpace(raw)
	if utf8.RuneCountInString(name) <= maxExternalSourceNameCodePoints {
		return name, false
	}
	runes := []rune(name)
	return string(runes[:maxExternalSourceNameCodePoints-1]) + "…", true
}

func normalizeRemoteProfileTitle(raw string) (string, []string) {
	name, shortened := normalizeRemoteExternalSourceName(raw)
	if !shortened {
		return name, nil
	}
	return name, []string{"Remote profile title exceeded 64 characters and was shortened."}
}

File: frontend/src/lib/externalSourceName.ts
export const MAX_EXTERNAL_SOURCE_NAME_CODE_POINTS = 64;

// The HTML maxLength property counts UTF-16 code units, so this allows the
// longest valid 64-code-point name while validation below enforces the API limit.
export const EXTERNAL_SOURCE_NAME_INPUT_MAX_LENGTH = MAX_EXTERNAL_SOURCE_NAME_CODE_POINTS * 2;

export function validateExternalSourceName(value: string): string | null {
  const name = value.trim();
  if (!name) return "Name is required";
  if (Array.from(name).length > MAX_EXTERNAL_SOURCE_NAME_CODE_POINTS) {
    return `Name is too long (max ${MAX_EXTERNAL_SOURCE_NAME_CODE_POINTS} characters)`;
  }
  return null;
}
